tool/parser.dart

```diff
@@ -0,0 +1,378 @@
+import 'package:charcode/charcode.dart';
+import 'package:quiver/core.dart';
+import 'package:quiver/collection.dart';
+
+/// A proof of concept rewriter for simple XML like expressions.
+/// 
+/// Technically all of these methods should be private but I think
+/// underscores are ugly.
+///
+///     new Dx('<div>${_foo}</div>'); =>
+///       new div()([_foo]);
+///
+///     new Dx('''
+///     <myElement bar=${fizz} bazz={2}>
+///       <div>Foo</div>
+///     </myElement>
+///     '''); =>
+///
+///      new MyElement(bar=fizz, bazz=2)([
+///        new Div()(["Foo"]),
+///      ]);
+///
+///      new Dx('''
+///        <div>
+///           "This is a text Ast"
+///           "${localTextAst}"
+///        </div>
+///      '''); =>
+///
+///      div()([text("this is a text Ast"), text(localTextAst)])
+///
+///      new Dx('<my-element></my-element>'); =>
+///
+///      new Element('my-element')();
+///
+class Parser {
+  List<Ast> _nodes;
+  List<int> _runes;
+  int _index;
+
+  Parser();
+
+  int get peek => _runes[_index];
+
+  /// Parses the given dart extension (dx) fragment into dart code.
+  Ast parse(String source) {
+    _nodes = <Ast>[];
+    _runes = source.runes.toList();
+    _index = 0;
+    consumeWhitespace();
```

Comment:
Do you need to `consumeWhitespace()` here if the first thing inside `while(true)` already does it?